### ISSUE TITLE
Compiled softmax/log-softmax with a non-power-of-two scale returns NaN for finite inputs (23 measured sites)

### ISSUE THREAD

**Issue by kenjohnscreates** [SlopDotCash/asi#2886]
Under `jax.jit` on XLA:CPU, `softmax(x / T)` and `log_softmax(x / T)` with a **non-power-of-two** scale return NaN for **finite** inputs once `|x / T|` reaches about 2^31. Eager evaluation of the same expression on the same inputs returns the correct result. I audited every softmax-family site in the package and measured **23 sites** across 7 core modules where this is reachable through the public API. Consequences range from a NaN policy handed back to the caller, through a wrong sampled action, to valid transitions being permanently rejected.

## Mechanism

XLA recomputes the product `x * c` (with `c = 1/T`) inside both fusions that the softmax lowers to: the reduce that computes `max(x*c)`, and the subtract-exponential that computes `exp(x*c - max)`. LLVM then contracts `x*c - max` into a single FMA (`fnmsub` on arm64). At the argmax the difference is therefore the **rounding error of the product**, about `|x/T| * 2^-24`, instead of exactly zero. Once that residual passes about 88 the exponential overflows or underflows and the normalisation becomes `inf/inf` or `0/0`.

Measured onset: the first NaN appears when `|x / T| >= 2^31`, where half an ulp is 128. `T = 1` (no multiply) and power-of-two `T` (exact multiply) are immune. `jax.lax.optimization_barrier` on the quotient is dropped from the HLO and does not help; a traced `T` and an explicit `x * float32(1/T)` behave identically.

**Immune construction — corrected, see the update at the end.** Subtract a maximum on *both* sides of the scale:

```python
centered = logits - jax.lax.stop_gradient(jnp.max(logits))
scaled = centered / temperature
return jax.nn.softmax(scaled - jax.lax.stop_gradient(jnp.max(scaled)))   # same for log_softmax
```

Softmax, log-softmax and cross entropy are all shift invariant, so both subtractions are exact rather than approximations. Neither one alone is sufficient: each blocks a different recomputation, and each was measured to fail on some shape in this repository. Details in the update below. It is not bitwise free: ordinary rows move by one or two ulp, because compiled main on those rows is itself the contraction artifact (compiled main already differs from eager on roughly 23 of 60 ordinary random rows at T in {0.3, 0.7, 0.99}).

## Measured sites

Every row below was measured on `f3d32c451ed1c1715e477ad56b782fc7ea89b206` by one agent and independently reproduced by a second that was asked to refute it. "Reached from init" means no parameter injection: only public calls with inputs the module's own validation accepts.

**Returns NaN to the caller**

| site | cheapest reachable path measured |
|---|---|
| `core/behavior_model.py:695` `predict_probabilities` | T=0.7, default step size: `init(1)` → `update(obs=[3e5], a=0)` (applied) → probabilities `[nan, nan]`; eager finite |
| `core/horde_actor_critic.py:628` `QHordeActorCritic.policy` | T=0.7: `init` → `start([1])` → `update(R=3e10)` (applied) → `[nan, nan]`; eager `[1, 0]` |
| `core/horde_actor_critic.py:924` `HordeActorCritic.policy` | T=0.7: `init` → `start([1])` → `update(R=1e10)` (applied) → `[nan, nan]`; eager `[0, 1]` |
| `core/recurrent_trace_actor_critic.py:2036` `policy` | `temperature=1e-12` **passes validation**: `init(4)` → `start` → `[nan, nan]` straight away; eager `[0, 1]` |

**Wrong action sampled**

| site | measured |
|---|---|
| `core/actor_critic.py:665` `ActorCriticAgent.policy` | `jr.categorical` takes `argmax(gumbel + log(probs))`, so a NaN wins the argmax: action 0 on 8/8 keys where eager picks 1 (see #2865) |
| `core/horde_actor_critic.py:644`, `:951` `select_action` | same, 8/8 keys after an applied finite update |
| `core/horde_actor_critic.py:1755`, `:2464` nonlinear policies | same, through `policy`/`select_action`/`start` after one applied update |
| `core/behavior_model.py:964` `sample_action` | the floor-and-renormalise step turns the NaN row into a uniform `[0.5, 0.5]`, so the learned policy is silently replaced by a coin flip rather than erroring |

**Valid transitions rejected (the finiteness gates work, and reject what they should accept)**

`core/behavior_model.py:835`, `core/horde_actor_critic.py:681`, `:1008`, `:1212` (the scan runner), `core/recurrent_trace_actor_critic.py:2057`, `core/options.py:1065`.

Two of these become persistent: `horde_actor_critic.py:763` (10 consecutive updates rejected, state bitwise unchanged each time) and `options.py:1065`, where a STOMP agent with `observation_dim=1` and a finite observation of 2.1e13 produces a NaN importance ratio, every following update is dropped (7 of 7 measured) and the agent livelocks while eager keeps learning. `options.py` needs **no non-default temperature at all**: its scale is the fixed `_GUMBEL_TIE_BREAK_SCALE = 1e-6`, and with a length-one contraction XLA turns the dot into an elementwise multiply that is recomputed in the fusion, so the existing `q - max(q)` centering is undone.

**Gradient dropped:** `core/behavior_model.py:763`, `core/horde_actor_critic.py:1306`, `:1339` (both inside `jax.grad`), `core/recurrent_trace_actor_critic.py:2282`.

**Wrong finite value committed, no gate:** `core/associative_memory.py` commits a shifted feature utility because the same contraction moves a cross-entropy term. That one is a wrong number rather than a NaN, so nothing rejects it; I filed #2885 for it separately.

## What is not affected

Every default temperature in the package is 1.0 or 0.5, both powers of two, so a default-configured agent is immune except at `options.py:1065`. Measured immune: `working_memory.py:560` and `delight.py:2047` (sigmoid has no max-subtraction), `resource_manager.py:531`/`:1049`, `compositional_features.py:705`, `feature_discovery.py:1009`, `prototype_memory.py:293`/`:440`, and the remaining `associative_memory.py` softmax helpers. In the benchmark lanes, the two IPMNIST screening vote sites were measured unreachable: the RLS vote temperature is 4.0, and the naive-Bayes and ensemble scales cannot reach the threshold from real inputs. I am deliberately not making a claim about `benchmarks/rule_discovery.py:563` here; it was flagged as possibly reachable and its independent re-measurement is still running, so it is excluded from the table above.

## Notes for whoever fixes this

- `core/average_reward.py` and `core/options.py` are **registered evidence sources**. Editing either invalidates its pinned claim until the frozen protocol is rerun, so those two want a deliberate decision rather than a drive-by patch. `average_reward.py:872` is also the one site measured to commit NaN into persistent state, via `start()`, which has no finiteness gate; the agent then wedges because every later update is rejected.
- `core/actor_critic.py:665` is already the subject of #2865. Its current head detects only an overflowing quotient, so it does not fix the reachable path at its own site; I posted the measured rows there.
- Tests that pin the compiled bits of `softmax(logits / T)` pin the contraction artifact. Comparing against eager, or restricting such assertions to power-of-two temperatures, pins the property instead.
- A per-module fix is probably better than one sweeping PR: ten open PRs currently touch these files.

## Method and limits

Both trees were run in isolated worktrees with import provenance asserted, under `XLA_FLAGS='--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=1'` and `OMP_NUM_THREADS=1`. Each claim was produced by one agent and independently re-measured by a second instructed to refute it; only claims that survived are listed. No benchmark, no seeds and no pinned artifact were run or touched, and nothing here is a performance or scientific-evidence claim. This is a correctness report on compiled float32 arithmetic.

Measured with claude-code / anthropic / claude-opus-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

---

## Update: the construction above is corrected, and four things were added

I kept the audit running after filing this. A follow-up round found a counterexample to the single-sided fix I first recommended, so the construction at the top of this issue has been corrected in place. Everything below is measured on the same host and toolchain.

**Neither centering alone is sufficient.** The two forms fail under different recomputations.

| construction | `options.py:1065` shape (length-one contraction, no bias add), 864 stress rows | `associative_memory.py` cross entropy |
|---|---|---|
| as written today | 621 compiled-only NaN | 5 of 11 regression rows fail |
| center the operand, then scale | **85 compiled-only NaN** | passes |
| scale, then center the product | 0 | **5 of 11 rows fail** |
| both | 0 | passes |

`options.py:1065` is the diagnostic case: it is *already written* as `softmax((q - max q) / 1e-6)` and is still defective, because `_q_values_for_obs` is a bare `q_weights @ observation` with `observation_dim=1` and no bias, so the multiply is the immediate producer of the max-subtraction and XLA re-forms it. Centering the operand only survives when that operand is materialised. Centering the product only survives when the subtraction itself is not re-formed. Applying both was immune everywhere it was tried.

**Regression rows must be unbatched.** A batch of 4 or more was measured to mask the defect entirely: identical rows that give 24 of 32 compiled-only NaN under per-row `jit` give zero under `jit(vmap(...))`. A natural batched parametrised sweep therefore passes on a broken build. No source-level predicate for this was found, so batched probes should be treated as unreliable for this class rather than as evidence of immunity.

**The active IPMNIST measurement path is clean.** A follow-up enumerated the screening and UPGD loss and mixture sites that were not in the first sweep: `ipmnist_screening.py` at 554, 904, 2121, 2616, 2631, 3292, 4435, 4458, 5536, 5537, 5543, 5545, 5553 and 5561, and `upgd_ipmnist.py:596`. All measured immune, for a consistent reason: the logits are a materialised dot, so both softmax fusions read the same buffer and there is no multiply to contract. Where a scale exists it is applied before a matmul, or it is a power of two. One development-only benchmark site, `rule_discovery.py:563`, returns a silently wrong, batch-size-dependent accuracy; that lane binds no source hash and is nonpromoting.

**Two sites are worth excluding from any fix PR**, as noted above: `average_reward.py:872` and `options.py:1065` are registered evidence sources, and both are low severity in reachability terms. `average_reward.py:872` needs `logit_clip` around 1e9 or a temperature below 1e-8; `options.py:1065` needs four simultaneous non-default settings. Trading a pinned artifact for either is a poor deal, and the diagnostic value of `options.py:1065` is in this issue rather than in a patch.

Also worth recording honestly: every measurement here is from one host and one toolchain, JAX 0.11.0 on XLA:CPU, arm64, single-threaded. The mechanism is backend instruction selection, so the specific thresholds and the immunity verdicts are properties of that configuration. The corrected construction is exact arithmetic either way, so it cannot be wrong elsewhere, only unnecessary.

#2885 fixes the associative-memory site with the corrected both-sided construction.

Measured with claude-code / anthropic / claude-opus-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by hermesagent270-commits** [comment]
I am starting a bounded, non-overlapping `core/horde_actor_critic.py` slice on current `main`: route the two linear and two nonlinear policy/gradient sites through one shift-invariant temperature-scaling helper, add public-path regressions for finite-policy/action/update behavior, and synthetic-merge the result with open PRs #2867–#2869 before submission. This module is not in the five-claim registered-source inventory; this is correctness-only development work with no performance or scientific claim.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Attribution status: self-reported
— [contributor]

**Comment by hermesagent270-commits** [comment]
Opened #2891 for the bounded Horde actor-critic slice. It routes both linear policies, both nonlinear policies, and both differentiated nonlinear policy objectives through one stable temperature scaler. Failing-first: all 6 new public regressions failed on exact `main`; candidate: 6/6 pass. The retained module panel passes 103 tests, synthetic composition with #2867–#2869 passes 171 tests, strict mypy and Ruff pass, and exact-head GitHub Actions run 34698196559 passed all 55 jobs.

This remains a correctness-only slice with no benchmark, performance, scientific, or evidence-promotion claim.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Attribution status: self-reported
— [contributor]

**Comment by hermesagent270-commits** [comment]
I am starting the remaining bounded `benchmarks/rule_discovery.py` slice on current `main`: stabilize the development-only naive-Bayes ensemble vote at its non-power-of-two input-dimension scale and add an unbatched compiled public-`rule_step` regression. I will not run a search or modify any output artifact; this lane is permanently nonpromoting and the module is outside the five-claim registered-source inventory.\n\nAI provider/model: openai / gpt-5\nClient / agent tooling: codex\nAttribution status: self-reported\n— [contributor]

**Comment by hermesagent270-commits** [comment]
Opened #2892 for the remaining development-only `rule_discovery.py` site. The NB vote now centers both before and after its non-power-of-two input-dimension scale. A finite, unbatched public `rule_step` regression fails on exact `main` with the wrong class and passes with the fix.

The retained rule-discovery and hostile-validation panel passes 94 tests; strict mypy and Ruff pass. Exact-head GitHub Actions run 34699289665 passed all 55 jobs, including macOS and Windows portability.

No benchmark or seed was run, no output artifact was touched, and this makes no performance, scientific, or evidence-promotion claim.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Attribution status: self-reported
